crates/paimon/src/spec/types.rs

```diff
@@ -0,0 +1,1035 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize, Hash)]
+pub struct DataType {
+    is_nullable: bool,
+    type_root: DataTypeRoot,
+}
+
+impl Display for DataType {
+    fn fmt(&self, _f: &mut Formatter<'_>) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+impl FromStr for DataType {
+    type Err = Error;
+
+    fn from_str(_: &str) -> Result<Self, Self::Err> {
+        todo!()
+    }
+}
+
+#[allow(dead_code)]
+impl DataType {
+    fn new(is_nullable: bool, type_root: DataTypeRoot) -> Self {
+        Self {
+            is_nullable,
+            type_root,
+        }
+    }
+
+    /// Returns true if the data type is nullable.
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L59>
+    fn is_nullable(&self) -> bool {
+        self.is_nullable
+    }
+
+    /// Returns the root of the data type.
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L66>
+    fn type_root(&self) -> &DataTypeRoot {
+        &self.type_root
+    }
+
+    /// Returns whether the root of the type equals to the type_root or not.
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L75>
+    fn is(&self, type_root: &DataTypeRoot) -> bool {
+        &self.type_root == type_root
+    }
+
+    /// Returns whether the family type of the type equals to the family or not.
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L103>
+    fn with_family(&self, family: DataTypeFamily) -> bool {
+        self.type_root.families().contains(family)
+    }
+
+    /// Returns whether the root of the type equals to at least on of the type_roots or not.
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L84>
+    fn is_any_of(&self, type_roots: &[DataTypeRoot]) -> bool {
+        type_roots.iter().any(|tr: &DataTypeRoot| self.is(tr))
+    }
+
+    /// Returns whether the root of the type is part of at least one family of the families or not.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L94>
+    fn is_any_with_family(&self, families: &[DataTypeFamily]) -> bool {
+        families
+            .iter()
+            .any(|f: &DataTypeFamily| self.with_family(f.clone()))
```

Comment:
`DataTypeFamily` is Copy, so I'm guessing we don't need to clone it.

Have you tried:

```rust
families
    .iter()
    .any(self.is_family)
```